### PR TITLE
HPCC-21981 Improve complete-uninstall to remove package hpccsystems-platform-with-spark.

### PR DESCRIPTION
**Pull request by gfortil**
Signed-off-by: Godson Fortil <Fortil.Godson@Lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:

Ubuntu 18.04:
Installed hpccsystems-platform
Applied the changes on complete-uninstall.sh
Completely uninstalled hpccsystems-platform: sudo /opt/HPCCSystems/sbin/complete-uninstall.sh
Copied complete-uninstall.sh contents into a new file with the same name
Ran complete-uninstall.sh (Package 'hpccsystems-platform' is not installed, so not removed)

Installed hpccsystems-platform-with-spark 
Applied the changes on complete-uninstall.sh
Completely uninstalled hpccsystems-platform-with-spark: sudo /opt/HPCCSystems/sbin/complete-uninstall.sh
Copied complete-uninstall.sh contents into a new file with the same name
Ran complete-uninstall.sh (Error, unable to find any hpcc package for removal.)

CentOS7:
Ubuntu 18.04:
Installed hpccsystems-platform
Applied the changes on complete-uninstall.sh
Completely uninstalled hpccsystems-platform: sudo /opt/HPCCSystems/sbin/complete-uninstall.sh
Copied complete-uninstall.sh contents into a new file with the same name
Ran complete-uninstall.sh (Error, unable to find any hpcc package for removal.)

Installed hpccsystems-platform-with-spark
Applied the changes on complete-uninstall.sh
Completely uninstalled hpccsystems-platform-with-spark: sudo /opt/HPCCSystems/sbin/complete-uninstall.sh
Copied complete-uninstall.sh contents into a new file with the same name
Ran complete-uninstall.sh (Error, unable to find any hpcc package for removal.)

### CODE REVIEW

**Comment by gfortil** [comment]
@mckellyln please review

File: initfiles/sbin/complete-uninstall.sh.in [hunk 1]
```diff
@@ -100,21 +100,35 @@ if [ -n "${lpath}" -a -h "${lpath}" ] ; then
...
+        echo "Removing ${pckname}"
+        yum remove -y ${pckname}
+        if [ $? -ne 0 ]; then
+           echo "An error has occured during removal.  Try to fix any issues and then run \`yum-complete-transaction\`."
+           exit 1
+        fi
     fi
 fi
 
+if [ -z "${pckname}" -a $force -eq 0] ; then
```
Comment (mckellyln) [inline initfiles/sbin/complete-uninstall.sh.in]: need a space after the 0 and before the ]

Comment (gfortil) [reply initfiles/sbin/complete-uninstall.sh.in]: Yes, I do. Thanks for pointing that out. I'll make the change.

**Comment by mckellyln** [comment]
@gfortil - look good - but please update the testing section above with what you tested.

File: initfiles/sbin/complete-uninstall.sh.in [hunk 2]
```diff
@@ -100,21 +100,35 @@ if [ -n "${lpath}" -a -h "${lpath}" ] ; then
...
 if [ -e /etc/debian_version ]; then
-    echo "Removing DEB"
-    apt-get remove -y hpccsystems-platform
-    if [ $? -ne 0 ]; then
-       echo "An error has occured during removal.  Try \`apt-get -f remove hpccsystems-platform\` to fix this issue."
-       exit 1
+    pckname=$(apt-cache pkgnames hpccsystems-platform 2> /dev/null)
+    if [ -n "${pckname}" ] ; then
+        echo "Removing ${pckname}"
+        apt-get remove -y ${pckname}
```
Comment (gfortil) [inline initfiles/sbin/complete-uninstall.sh.in]: need to purge to remove package name left behind after removal
dpkg -P ${pckname}

File: initfiles/sbin/complete-uninstall.sh.in [hunk 3]
```diff
@@ -100,21 +100,36 @@ if [ -n "${lpath}" -a -h "${lpath}" ] ; then
...
+        dpkg -P ${pckname}
+        if [ $? -ne 0 ]; then
```
Comment (mckellyln) [inline initfiles/sbin/complete-uninstall.sh.in]: I think you want to do the dpkg -P only if the apt-get remove was successful, so you need to move the dpkg cmd to after the if test.
```
apt-get remove -y ${pckname}
if [ $? -ne 0 ] ; then
    echo "Error ..."
    exit
fi
dpkg -P ${pckname}
```

Comment (mckellyln) [reply initfiles/sbin/complete-uninstall.sh.in]: Not sure but maybe a single cmd of
```
apt-get purge ${pckname}
```
would work ?  Just a thought tho, I am not sure which is best

Comment (gfortil) [reply initfiles/sbin/complete-uninstall.sh.in]: I don't think there is any difference between dpkg and apt when it comes to uninstalling. They go through the same process of purging the package.

Comment (mckellyln) [reply initfiles/sbin/complete-uninstall.sh.in]: Right, but $? is the status of the most recent command, so if the apt-get remove fails - do you want to issue the dpkg -P ?
If leaveenv is set on the command-line then we leave the config files - and does a purge remove those ?  If so we dont want to do a purge.
So lets go back to before - why do we need the purge command ?

**Comment by mckellyln** [comment]
I am not a fan of the purge.
Perhaps this is the better logic -
```
    pckname=$(apt-cache pkgnames hpccsystems-platform 2> /dev/null)
    if [ -n "${pckname}" ] ; then
        echo "Removing ${pckname}"
        apt-get remove -y ${pckname}
        if [ $? -ne 0 -a $force -eq 0 ]; then
           echo "An error has occured during removal.  Try \`apt-get -f remove hpccsystems-platform\` to fix this issue."
           exit 1
        fi
    fi
```
Then it is like before and if there is a failure to remove the pkg you can rerun and go on if you specify the -force on the cmdline ? 
What do you think ?

**Review by mckellyln — APPROVED** [review]

**Comment by mckellyln** [comment]
@gfortil looks good.

**Pull request merged**